### DISCUSSION TITLE
[Share] 根据注释颜色，添加标签

### DISCUSSION THREAD

**Discussion by appwcn** [windingwind/zotero-actions-tags#339] [Action Scripts]
### Description
根据注释颜色，更新注释标签。（使用非默认颜色，需结合zotery style插件）
正确代码已由windingwind大神更新，移步以下链接复制：
https://github.com/windingwind/zotero-actions-tags/discussions/339#discussioncomment-10046647

### Event

Creat Annotation

### Operation

Script

### Data

```javascript
/**
 * 注释：根据注释颜色增加标签
 * @author appwcn
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/339
 */

if (!item) {
  return;
}

let topItem = Zotero.Items.getTopLevel([item])[0];

return await EditAnnotation(item);

async function EditAnnotation(annotationItem) {
  if (!annotationItem.isAnnotation()) {
    return;
  }
  if (annotationItem.annotationColor == null) {
    return "未读取到注释颜色";
  }
  // Fixed: Added comma between parameters
  annotationItem.annotationComment= await FormatText(annotationItem.annotationColor, annotationItem.annotationComment);
  return "更新文献密码注释标签";
}

async function FormatText(color, text) {
  let tag = "";
  tag = color === "#ffd400" ? " #文献密码/研究意义" :
        color === "#ff6666" ? " #文献密码/研究问题" :
        color === "#5fb236" ? " #文献密码/理论依据" :
        color === "#2ea8e5" ? " #文献密码/研究结论" :
        color === "#e56eee" ? " #文献密码/文献综合" :
        color === "#f19837" ? " #文献密码/研究技术" : tag;
  if (!text) {
	return tag;
  } else {
	return text + "\n" + tag;
  }
}
```

**Comment by windingwind** [comment]
Tag annotation by color:

<img width="373" alt="image" src="https://github.com/user-attachments/assets/3d230f2b-b46c-46f6-a65f-a858669edc9b">


```js
/**
 * 注释：根据注释颜色增加标签
 * @author appwcn windingwind
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/339
 */

// EDTI TAG MAPPING BELOW
const tags = {
	"#ffd400": "*TAG1",
	"#ff6666": "*TAG2",
	"#5fb236": "*TAG3",
	"#2ea8e5": "*TAG4",
	"#e56eee": "*TAG5",
	"#f19837": "*TAG6",
}

if (!item) {
	return;
}

if (!item.isAnnotation() || !item.annotationColor) {
	return;
}

const tag = tags[item.annotationColor];
item.addTag(tag);
return `Tag added: ${tag}`;
```

**Reply by Dobby2046** [reply]
在添加新的颜色后，运行提示：Script Error: Tag data must contain 'tag' property

**Reply by windingwind** [reply]
若修改了颜色，需要对应修改`// EDTI TAG MAPPING BELOW`一行下方的颜色映射，保证你每个颜色都有对应的标签

**Reply by Fan-tas-tic** [reply]
一一对应设置了映射，但只有一个颜色能自动添加标签，其他的不可以，请问是什么问题嘛？

/**
 * 注释：根据注释颜色增加标签
 * @author appwcn windingwind
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/339
 */

// EDTI TAG MAPPING BELOW
const tags = {
	"#F9F871": "#方法",
	"#ff8066": "#可用词句",
	"#5fb236": "#综述背景",
	"#2ea8e5": "#结论",
	"#4FFBDF": "#创新",
	"#2F7AFF": "#展望",
    "#3596B5": "#词句概念",
    "#C2B280": "#不足",
}

if (!item) {
	return;
}

if (!item.isAnnotation() || !item.annotationColor) {
	return;
}

const tag = tags[item.annotationColor];
item.addTag(tag);
return `Tag added: ${tag}`;

**Reply by Fan-tas-tic** [reply]
问题搞定了，颜色代码得小写！

**Reply by llity** [reply]
发现一个问题，注释颜色更改后，tag名称没有改变。

**Reply by fredericky123** [reply]
> 问题搞定了，颜色代码得小写！

同吃过这个亏

**Reply by Neokettle** [reply]
> 发现一个问题，注释颜色更改后，tag名称没有改变。

是有这个问题

**Reply by okerivy** [reply]
修改了下代码

- 添加注释，可以自动添加对应的标签🏷️
- 注释颜色更改后，标签并不会改变，但是 可以**右键菜单更新标签**
- 注意：**_代码中，请更换成自己的颜色和标签_**


<img width="266" alt="image" src="https://github.com/user-attachments/assets/1f3e6538-ba99-42b8-acd8-0306241eeefd">


``` 
/**
 * 注释：根据注释颜色更新标签（移除旧标签并添加新标签）
 * @description 通过注释颜色自动更新标签，包括移除旧标签并添加新标签
 * 
 * @author appwcn windingwind
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/339
 */

// 定义颜色与标签的映射（请更换成自己的颜色和标签）
const tags = {
    "#ffd400": "结论",
    "#0033ff": "摘录",
    "#ff4349": "原因"
};


const allTags = Object.values(tags);


if (!item) {
    return "No item provided.";
}

if (!item.isAnnotation() || !item.annotationColor) {
    return "This item is not an annotation or has no color.";
}


const annotationColor = item.annotationColor.toLowerCase();
const newTag = tags[annotationColor];


const title = item.getField('title');


const existingTags = item.getTags().map(tag => tag.tag); 
let removedTags = [];


for (const tag of existingTags) {
    if (allTags.includes(tag)) {
        item.removeTag(tag);
        removedTags.push(tag);
    }
}


if (newTag) {
    item.addTag(newTag);
    return `Title: "${title}" | Removed: [${removedTags.join(", ")}] | Added: ${newTag}`;
} else if (removedTags.length > 0) {
    return `Title: "${title}" | Removed: [${removedTags.join(", ")}] | No new tag added.`;
} else {
    return `Title: "${title}" | No matching tag for color: ${annotationColor}.`;
}

 ```

**Reply by Archetto-O** [reply]
我的情况下会同时添加带# 和不带#的两个同样标签。在addTag之后加上了一段判断之后可以解决。ai写的代码，不知是否有更好方案。
```
if (newTag) {
    item.addTag(newTag);

    const tags = item.getTags().map(tag => tag.tag);
    const tagsToRemove = tags.filter(tag => !tag.startsWith('#'));
    tagsToRemove.forEach(tag => {item.removeTag(tag);});

    return `Title: "${title}" | Removed: [${removedTags.join(", ")}] | Added: ${newTag}`;
```

**Reply by zhouguangpei313313** [reply]
<img width="298" height="259" alt="image" src="https://github.com/user-attachments/assets/49b562a0-ff2e-4713-a9f5-8b0db4f5cf78" />
请问不需要这个，怎么修改代码

**Comment by luckydoudoudou** [comment]
如何根据注释的颜色标签，自动创建相应的标签的笔记内容

**Reply by luckydoudoudou** [reply]
如何根据注释的颜色标签，自动创建相应的标签的笔记内容，求代码

**Reply by 7stella** [reply]
b站有教程
https://www.bilibili.com/video/BV1paWeepEfW/?spm_id_from=333.337.search-card.all.click&vd_source=bcb6f47a6902141379eb69fb3a99873d

**Comment by YYL924** [comment]
![image](https://github.com/user-attachments/assets/562f1ef1-6637-48e8-adb1-b0e139828581)
![屏幕截图 2024-10-01 003134](https://github.com/user-attachments/assets/c3f77659-8ec6-4526-b024-2276577bc95d)


/**
 * 注释：根据注释颜色增加标签
 * @author appwcn windingwind
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/339
 */

// EDTI TAG MAPPING BELOW
const tags = {
	"#FC8C8C": "*TAG表",
	"#F7A072": "*TAG图",
	"#FFE066": "*TAG重点",
	"#78C0E0": "*TAG不懂",
	"#D1AEEF": "*TAG抄",
	"#f19837": "*TAG标题基本结构",
}

if (!item) {
	return;
}

if (!item.isAnnotation() || !item.annotationColor) {
	return;
}

const tag = tags[item.annotationColor];
item.addTag(tag);
return `Tag added: ${tag}`;

**Reply by Fan-tas-tic** [reply]
颜色代码得小写！

**Comment by ZiHao256** [comment]
Select tags parsed from the Zotero Style color group and add it to the annotation. 
I refer to #339 [comment](https://github.com/windingwind/zotero-actions-tags/discussions/339#discussioncomment-10046647) and #108.

```
/**
 * Select tags parsed from the Zotero Style color group and add it to the annotation. 
 */
const window = require("window");
const Zotero = require("Zotero");

function selectItems(itemList) {
    var io = { dataIn: itemList, dataOut: null };
    window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
        "_blank", "chrome,modal,centerscreen,resizable=yes", io);
    return io.dataOut;
}

function getTagsByColor(colorCode) {
	const rawData = JSON.parse(Zotero.Prefs.get('zoterostyle.annotationColors'));
    for (let item of rawData) {
        if (item["1"] === colorCode) {
            return item["0"].split('/');
        }
    }
}

if (!item) {
    return;
}

if (!item.isAnnotation() || !item.annotationColor) {
    return;
}

const tags = getTagsByColor(item.annotationColor);

const selectedTags = selectItems(tags);

if (selectedTags) {
	for (const tag of Object.values(selectedTags)) {
		item.addTag(tag);
	}
	return `Tags added: ${Object.values(selectedTags).join(", ")}`;
}
```
